src/V3SchedTiming.cpp

```diff
@@ -0,0 +1,338 @@
+                        awaitp->statement(true);
+                        m_forkp->addHereThisAsNext(awaitp);
+                    }
+                } else {
+                    refp->v3warn(E_UNSUPPORTED, "Unsupported: variable local to a forking process "
+                                                "accessed in a fork..join_any or fork..join_none");
+                    return;
+                }
+                // Remap the reference
+                AstVarScope* const vscp = refp->varScopep();
+                if (!vscp->user2p()) {
+                    // Clone the var to the new function
+                    AstVar* const varp = refp->varp();
+                    AstVar* const newvarp
+                        = new AstVar{varp->fileline(), VVarType::BLOCKTEMP, varp->name(), varp};
+                    newvarp->funcLocal(true);
+                    newvarp->direction(passByValue ? VDirection::INPUT : VDirection::REF);
+                    funcp->addArgsp(newvarp);
+                    AstVarScope* const newvscp
+                        = new AstVarScope{newvarp->fileline(), funcp->scopep(), newvarp};
+                    funcp->scopep()->addVarp(newvscp);
+                    vscp->user2p(newvscp);
+                    callp->addArgsp(new AstVarRef{refp->fileline(), vscp, VAccess::READ});
+                }
+                auto* const newvscp = VN_AS(vscp->user2p(), VarScope);
+                refp->varScopep(newvscp);
+                refp->varp(newvscp->varp());
+            });
+        }
+
+        // VISITORS
+        virtual void visit(AstNodeModule* nodep) override {
+            VL_RESTORER(m_inClass);
+            m_inClass = VN_IS(nodep, Class);
+            iterateChildren(nodep);
+        }
+        virtual void visit(AstCFunc* nodep) override {
+            m_funcp = nodep;
+            iterateChildren(nodep);
+            m_funcp = nullptr;
+        }
+        virtual void visit(AstVar* nodep) override { nodep->user1(true); }
+        virtual void visit(AstFork* nodep) override {
+            VL_RESTORER(m_forkp);
+            m_forkp = nodep;
+            iterateChildrenConst(nodep);  // Const, so we don't iterate the calls twice
+            // Replace self with the function calls (no co_await, as we don't want the main process
+            // to suspend whenever any of the children do)
+            nodep->replaceWith(nodep->stmtsp()->unlinkFrBackWithNext());
+            VL_DO_DANGLING(nodep->deleteTree(), nodep);
+        }
+        virtual void visit(AstBegin* nodep) override {
+            UASSERT_OBJ(m_forkp, nodep, "Begin outside of a fork");
+            UASSERT_OBJ(!nodep->name().empty(), nodep, "Begin needs a name");
+            FileLine* const flp = nodep->fileline();
+            // Create a function to put this begin's statements in
+            AstCFunc* const newfuncp
+                = new AstCFunc{flp, nodep->name(), m_funcp->scopep(), "VlCoroutine"};
```

Comment:
CFunc names need to be unique per AstModule, otherwise we will emit a double definition, is this sufficient to ensure uniqueness?

Reply:
The begins get unique names in `V3Timing`, from a `V3UniqueNames`. I do it there so the names are consistent with the relevant fork sync vars.